strawberryfields/decompositions.py

```diff
@@ -400,44 +414,139 @@ def mach_zehnder(m, n, internal_phase, external_phase, nmax):
+def mach_zehnder_inv(m, n, phi_int, phi_ext, nmax):
+    r"""The inverse of the mach_zehnder matrix."""
+    return mach_zehnder(m, n, phi_int, phi_ext, nmax).conj().T
+
+
+def nullMZi(m, n, U):
+    r"""Nullifies element m,n of U using mach_zehnder_inv.
+
+    Args:
+        m (int): row index of element to be nullified
+        n (int): column index of element to be nullified
+        U (array): matrix whose m,n element is to be nullified
+
+    Returns:
+        list: list containing ``[m, n, internal_phase, external_phase, nmax]`` of the
+            mach_zehnder_inv unitaries needed
+    """
+    (nmax, mmax) = U.shape
+
+    if nmax != mmax:
+        raise ValueError("U must be a square matrix")
 
-    This decomposition starts with the output from :func:`rectangular_phase_end`
-    and further decomposes each of the T unitaries into Mach-Zehnder
-    interferometers consisting of two phase-shifters and two symmetric (50:50)
-    beamsplitters.
+    if U[m, n] == 0:
+        phi_i = np.pi
+        phi_e = 0
+    elif U[m, n + 1] == 0:
+        phi_i = 0
+        phi_e = 0
+    else:
+        r = -U[m, n + 1] / U[m, n]
+        phi_i = 2 * np.arctan(np.abs(r))
+        phi_e = -np.angle(r)
 
-    The two beamsplitters in this decomposition of T are modeled by :class:`~.ops.BSgate`
-    with arguments :math:`(\pi/4, \pi/2)`, and the two phase-shifters (see :class:`~.ops.Rgate`)
-    act on the input mode with the lower index of the two. The phase imposed
-    by the first phaseshifter (before the first beamsplitter) is named
-    ``external_phase``, while we call the phase shift between the beamsplitters
-    ``internal_phase``.
+    return [n, n + 1, phi_i, phi_e, nmax]
 
-    The algorithm applied in this function makes use of the following identity:
 
-    .. code-block:: python
+def nullMZ(n, m, U):
+    r"""Nullifies element n,m of U using mach_zehnder.
 
-        Rgate(alpha) | 1
-        Rgate(beta) | 2
-        Rgate(phi) | 1
-        BSgate(theta, 0) | 1, 2
+    Args:
+        n (int): row index of element to be nullified
+        m (int): column index of element to be nullified
+        U (array): matrix whose m,n element is to be nullified
 
-        equals
+    Returns:
+        list: list containing ``[m, n, internal_phase, external_phase, nmax]`` of the
+            mach_zehnder unitaries needed
+    """
+    (nmax, mmax) = U.shape
 
-        Rgate(phi+alpha-beta) | 1
-        BSgate(pi/4, pi/2) | 1, 2
-        Rgate(2*theta+pi) | 1, 2
-        BSgate(pi/4, pi/2) | 1, 2
-        Rgate(beta-theta+pi) | 1
-        Rgate(beta-theta) | 2
+    if nmax != mmax:
+        raise ValueError("U must be a square matrix")
+
+    if U[n, m] == 0:
+        phi_i = np.pi
+        phi_e = 0
+    elif U[n - 1, m] == 0:
+        phi_i = 0
+        phi_e = 0
+    else:
+        r = U[n - 1, m] / U[n, m]
+        phi_i = 2 * np.arctan(np.abs(r))
+        phi_e = -np.angle(r)
```

Comment:
Same here re: adding brief comments explaining each branch. Just a simple 1-liner is fine, just helps future developers if they need to track down bugs etc :)

Reply:
Sounds good!